lib/Parse/Lexer.cpp

```diff
@@ -932,20 +932,27 @@ void Lexer::lexHexNumber() {
   const char *TokStart = CurPtr-1;
   assert(*TokStart == '0' && "not a hex literal");
 
-  // 0x[0-9a-fA-F][0-9a-fA-F_]*
-  ++CurPtr;
-  if (!isHexDigit(*CurPtr)) {
-    diagnose(CurPtr, diag::lex_expected_digit_in_int_literal);
+  auto expected_hex_digit = [&]() {
+    diagnose(CurPtr, diag::lex_expected_digit_in_int_literal, 3 /*Hex*/);
```

Comment:
I think it would have to be available in all of the lexer since it would be used from two methods (`lexNumber` and `lexHexNumber`). Is that still preferable even if it's only used for a diagnostic?